internal/framework6provider/float64_precision_resource_test.go

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: MPL-2.0
+
+package framework
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-framework/providerserver"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/plancheck"
+)
+
+// This test verifies that the correct precision value is stored in state
+// https://github.com/hashicorp/terraform-plugin-framework/issues/815
+func TestSchemaResource_Float64Attribute_Precision(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		ProtoV6ProviderFactories: map[string]func() (tfprotov6.ProviderServer, error){
+			"framework": providerserver.NewProtocol6WithError(New()),
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: `resource "framework_float64_precision" "test" {
+					float64_attribute = 1 - 0.99
+				}`,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr("framework_float64_precision.test", "float64_attribute", "0.010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003"),
+				),
+			},
+			{
+				Config: `resource "framework_float64_precision" "test" {
+					float64_attribute = 1 - 0.98
+				}`,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr("framework_float64_precision.test", "float64_attribute", "0.020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006"),
+				),
+			},
+		},
+	})
+}
+
+// This test verifies that a plan is detected when encountering a more precise value
+// https://github.com/hashicorp/terraform-plugin-framework/issues/815
+func TestSchemaResource_Float64Attribute_Precision_Plan(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		ProtoV6ProviderFactories: map[string]func() (tfprotov6.ProviderServer, error){
+			"framework": providerserver.NewProtocol6WithError(New()),
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: `resource "framework_float64_precision" "test" {
+					float64_attribute = 0.01
+				}`,
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr("framework_float64_precision.test", "float64_attribute", "0.01"),
+				),
+			},
+			{
+				Config: `resource "framework_float64_precision" "test" {
+					float64_attribute = 1 - 0.99
+				}`,
+				ConfigPlanChecks: resource.ConfigPlanChecks{
+					PreApply: []plancheck.PlanCheck{
+						plancheck.ExpectNonEmptyPlan(),
+					},
+				},
+				Check: resource.ComposeAggregateTestCheckFunc(
+					resource.TestCheckResourceAttr("framework_float64_precision.test", "float64_attribute", "0.010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003"),
+				),
+			},
```

Comment:
Personally, I think Terraform's behavior here is quite unfortunate, but that's an upstream problem to handle.

Reply:
Indeed 😕